tests/subsys/bluetooth/fast_pair/storage/src/main.c

```diff
@@ -0,0 +1,312 @@
+
+	generate_account_key(test_key_id, account_key);
+	err = fp_storage_account_key_save(account_key);
+	zassert_ok(err, "Unexpected error");
+
+	/* Try to add key duplicate. */
+	err = fp_storage_account_key_save(account_key);
+	zassert_ok(err, "Unexpected error");
+
+	uint8_t read_keys[ACCOUNT_KEY_MAX_CNT][ACCOUNT_KEY_LEN];
+	size_t read_cnt = ACCOUNT_KEY_MAX_CNT;
+
+	zassert_equal(fp_storage_account_key_count(), 1, "Invalid Account Key count");
+	err = fp_storage_account_keys_get(read_keys, &read_cnt);
+	zassert_ok(err, "Getting Account Keys failed");
+	zassert_equal(read_cnt, 1, "Invalid Account Key count");
+	zassert_true(check_account_key_id(test_key_id, read_keys[0]), "Invalid key on read");
+
+	/* Reload keys from storage and validate them again. */
+	reload_keys_from_storage();
+	memset(read_keys, 0, sizeof(read_keys));
+	read_cnt = ACCOUNT_KEY_MAX_CNT;
+
+	zassert_equal(fp_storage_account_key_count(), 1, "Invalid Account Key count");
+	err = fp_storage_account_keys_get(read_keys, &read_cnt);
+	zassert_ok(err, "Getting Account Keys failed");
+	zassert_equal(read_cnt, 1, "Invalid Account Key count");
+	zassert_true(check_account_key_id(test_key_id, read_keys[0]), "Invalid key on read");
+}
+
+static void generate_and_store_keys(uint8_t gen_count)
+{
+	int err;
+	uint8_t account_key[ACCOUNT_KEY_LEN];
+
+	for (uint8_t i = 0; i < gen_count; i++) {
+		generate_account_key(i, account_key);
+
+		err = fp_storage_account_key_save(account_key);
+		zassert_ok(err, "Failed to store Account Key");
+	}
+}
+
+static void test_invalid_calls(void)
+{
+	static const uint8_t test_key_cnt = 3;
+
+	int err;
+	uint8_t found_key[ACCOUNT_KEY_LEN];
+	uint8_t read_keys[ACCOUNT_KEY_MAX_CNT][ACCOUNT_KEY_LEN];
+
+	generate_and_store_keys(test_key_cnt);
+
+	for (size_t read_cnt = 0; read_cnt <= ACCOUNT_KEY_MAX_CNT; read_cnt++) {
+		size_t read_cnt_ret = read_cnt;
+
+		err = fp_storage_account_keys_get(read_keys, &read_cnt_ret);
+
+		if (read_cnt < test_key_cnt) {
+			zassert_equal(err, -EINVAL, "Expected error when array is too small");
+		} else {
+			zassert_ok(err, "Unexpected error");
+			zassert_equal(read_cnt_ret, test_key_cnt, "Inavlid account key count");
+
+			for (size_t i = 0; i < read_cnt_ret; i++) {
+				zassert_true(check_account_key_id(i, read_keys[i]),
+					     "Invalid key on read");
+			}
+		}
+	}
+
+	uint8_t key_id = 0;
+
+	err = fp_storage_account_key_find(found_key, NULL, &key_id);
+	zassert_equal(err, -EINVAL, "Expected error when callback is NULL");
+}
+
+static void test_find(void)
+{
+	static const size_t test_key_cnt = 3;
+	int err = 0;
+
+	generate_and_store_keys(test_key_cnt);
+
+	for (uint8_t key_id = 0; key_id < test_key_cnt; key_id++) {
+		uint8_t account_key[ACCOUNT_KEY_LEN];
+		uint8_t prev_key_id = key_id;
+
+		err = fp_storage_account_key_find(account_key, account_key_find_cb, &key_id);
+		zassert_ok(err, "Failed to find Account Key");
+		zassert_equal(prev_key_id, key_id, "Context should not be modified");
```

Comment:
I think we should allow context modification to make this API more flexible.

Of course, the context modification can only happen in the callback logic and is not permitted to happen in the storage module logic.

I guess this is just an assertion for this specific callback type _account_key_find_cb_ which doesn't modify the context. We still allow context modification in other uses cases, don't we?

Reply:
Of course we allow context modification in general. We just know that the callback in the unit test does not modify the context so we can assert this here